numpy/_core/src/multiarray/dtype_transfer.c

```diff
@@ -3831,4 +3825,32 @@ PyArray_PrepareThreeRawArrayIter(int ndim, npy_intp const *shape,
     *out_dataC = dataC;
     *out_ndim = ndim;
     return 0;
-}
\ No newline at end of file
+}
+
+/*
+ * Could be made public API. Check the result and, if needed, FPE status after
+ * a call to a PyArrayMethod_StridedLoop function. If ret < -100, PyErr has
+ * already been set, even if NPY_METH_REQUIRES_PYAPI is not set.
+ */
+NPY_NO_EXPORT int
+PyArray_CheckRetAndFPE(const char * name, int ret, NPY_ARRAYMETHOD_FLAGS flags)
+{
+    int res = 0;
+    if (flags & NPY_METH_REQUIRES_PYAPI && PyErr_Occurred()) {
```

Comment:
As mentioned elsewhere this logic looks wrong.  It should probably just _assume_ `-1` means an error is already set.

Reply:
redone